mesa_geo/raster_layers.py

```diff
@@ -0,0 +1,476 @@
+from __future__ import annotations
+
+import copy
+import itertools
+import math
+from typing import (
+    Tuple,
+    List,
+    Dict,
+    Any,
+    Set,
+    cast,
+    overload,
+    Sequence,
+    Iterator,
+    Iterable,
+)
+
+import numpy as np
+import rasterio as rio
+from affine import Affine
+from mesa.agent import Agent
+from mesa.space import Coordinate, accept_tuple_argument
+from rasterio.warp import (
+    calculate_default_transform,
+    transform_bounds,
+    reproject,
+    Resampling,
+)
+
+from mesa_geo.geo_base import GeoBase
+
+
+class RasterBase(GeoBase):
+    """
+    Base class for raster layers.
+    """
+
+    _width: int
+    _height: int
+    _transform: Affine
+    _total_bounds: np.ndarray  # [min_x, min_y, max_x, max_y]
+
+    def __init__(self, width, height, crs, total_bounds):
+        super().__init__(crs)
+        self._width = width
+        self._height = height
+        self._total_bounds = total_bounds
+        self._update_transform()
+
+    @property
+    def width(self) -> int:
+        return self._width
+
+    @width.setter
+    def width(self, width: int) -> None:
+        self._width = width
+        self._update_transform()
+
+    @property
+    def height(self) -> int:
+        return self._height
+
+    @height.setter
+    def height(self, height: int) -> None:
+        self._height = height
+        self._update_transform()
+
+    @property
+    def total_bounds(self) -> np.ndarray:
+        return self._total_bounds
+
+    @total_bounds.setter
+    def total_bounds(self, total_bounds: np.ndarray) -> None:
+        self._total_bounds = total_bounds
+        self._update_transform()
+
+    @property
+    def transform(self) -> Affine:
+        return self._transform
+
+    @property
+    def resolution(self) -> Tuple[float, float]:
+        """
+        Returns the (width, height) of a cell in the units of CRS.
+        """
+        a, b, _, d, e, _, _, _, _ = self.transform
+        return math.sqrt(a**2 + d**2), math.sqrt(b**2 + e**2)
+
+    def _update_transform(self) -> None:
+        self._transform = rio.transform.from_bounds(
+            *self.total_bounds, width=self.width, height=self.height
+        )
+
+    def to_crs(self, crs, inplace=False) -> RasterBase | None:
+        raise NotImplementedError
+
+    def out_of_bounds(self, pos: Coordinate) -> bool:
+        x, y = pos
+        return x < 0 or x >= self.width or y < 0 or y >= self.height
+
+
+class Cell(Agent):
```

Comment:
What is the definition of `Cell`?

Reply:
Cells are building blocks of RasterLayer, and can have multiple attributes, depending on specific models. Some examples I wrote from the separated branch:

- Rainfall example: https://github.com/wang-boyu/mesa-geo/blob/2622e870784aa06b55d52cb9697f8caea65de558/examples/rainfall/rainfall/space.py#L11-L23
- Urban growth example: https://github.com/wang-boyu/mesa-geo/blob/2622e870784aa06b55d52cb9697f8caea65de558/examples/urban_growth/urban_growth/space.py#L10-L43

In theory, they can act like agents if the `step()` method is implemented, or stay static. But one thing I'm not sure (that I forgot to mention in the PR description) is how to make them work with the schedulers, which needs a reference/pointer to the `model` within the agent/cell.

Reply:
Ah, so it is a container for attributes. It seems that it is going to be slow for huge raster grid size. Makes me wonder that the raster layer can be reimplemented in Cython for faster computation. But this can be done in future iterations.